### DISCUSSION TITLE
setdata?

### DISCUSSION THREAD

**Discussion by HFolkertH** [cincheo/daquota#34] [Q&A] [answered]
hello Renaud,

The blog says that this should work:
![image](https://user-images.githubusercontent.com/22573455/170816338-2ba51706-bada-4159-a8e0-45cf5071b3db.png)

But as you can see I get an error.

also with target = text-16 and action = setdata(data) and argument = 'hello', I still get an error.

Text-16 is part of a dialog-0.

also does not work: 

![image](https://user-images.githubusercontent.com/22573455/170816438-cdf47dd0-414f-412e-8f3b-a2aeda048c3f.png)

They all work in the IDE but not in de run_browser. What am i doing wrong?

.

**Comment by renaudpawlak** [comment]
These were no "actual" errors. They were bugs in the code editor, so I have fixed them and pushed a new version. You can give it another try (don't forget to reload and check that the version is now 0.9.13).  

Note that ``$c('text-16').setData('hello')`` is correct and should work. On the other hand, ``$d('text-16').setData('hello')`` cannot work because ``$d('text-16')`` returns the data (not the component). An alternative is also ``$c('text-16').dataModel = 'hello'``.

Last but not least, the simplest way to do it in your case is to select ``text-16`` as a target and select the ``setData`` action in the action selector (then pass the argument you want, like ``'hello'``).

**Reply by HFolkertH** [reply]
Hello Renaud
I tried your last suggestion but somehow it does not work.

It's just a button that puts "hello" in text-0 ( which is part of dialog-0) and then shows dialog-0.
```
{
  "layout": "block",
  "components": [
    {
      "layout": "block",
      "components": [
        {
          "dataSource": "$parent",
          "label": "Click me",
          "buttonType": "button",
          "variant": "secondary",
          "pill": false,
          "squared": false,
          "block": false,
          "disabled": false,
          "eventHandlers": [
            {
              "global": false,
              "name": "@click",
              "actions": [
                {
                  "targetId": "text-0",
                  "name": "setData",
                  "description": "Default action",
                  "argument": "'hallo'\n"
                },
                {
                  "targetId": "dialog-0",
                  "name": "show",
                  "description": "New action",
                  "argument": "undefined"
                }
              ]
            }
          ],
          "type": "ButtonView",
          "cid": "button-0"
        }
      ],
      "defaultValue": "={}",
      "type": "ContainerView",
      "eventHandlers": [],
      "cid": "container-0"
    },
    {
      "title": "",
      "content": {
        "tag": "p",
        "text": "Lorem ipsum dolor sit amet.",
        "type": "TextView",
        "eventHandlers": [
          {
            "global": false,
            "name": "@click",
            "actions": [
              {
                "targetId": "text-0",
                "name": "setData",
                "description": "Default action",
                "argument": "'hallo'\n"
              },
              {
                "targetId": "dialog-0",
                "name": "show",
                "description": "New action",
                "argument": "undefined"
              }
            ]
          }
        ],
        "cid": "text-0",
        "dataSource": "$parent"
      },
      "type": "DialogView",
      "eventHandlers": [],
      "cid": "dialog-0",
      "dataSource": "$parent"
    }
  ],
  "defaultValue": "={}",
  "type": "ContainerView",
  "eventHandlers": [],
  "cid": "index"
}
```

 
BTW should the "cid" not be the first line of each component? 
 It makes code much more readable.

**Reply by renaudpawlak** [reply]
I see. That's because you are setting the data of a text component. Your ``setData`` works but you need to clear the value of ``text`` in the text component, because (as it says), it overrides the data value. Currently, your text is worth ``Lorem ipsum dolor sit amet.``... you need to remove it. You can also change it with a formula that depends on the ``dataModel``.

**Reply by HFolkertH** [reply]
Hello Renaud,

Thanks for explaining, but it just does not work. Can you provide a working example?
I have an empty text-field (no default "lorem...."  ) in a dialog. The button tries to setdata  this text-field to "hello"  but all I get is "invalid text".

**Reply by renaudpawlak** [reply]
Ok, my bad I did not test enough. You need to call setData when the dialog is opened, otherwise, the text component does not exist in the page and ``$c('text-0')`` will fail. It does not fail in edit mode because the dialog is always visible. So, what you need to do it to change the order of your button click actions to (1) open the dialog (2) set the data. I cut and paste the associated code:

```json
{
  "layout": "block",
  "components": [
    {
      "layout": "block",
      "components": [
        {
          "dataSource": "$parent",
          "label": "Click me",
          "buttonType": "button",
          "variant": "secondary",
          "pill": false,
          "squared": false,
          "block": false,
          "disabled": false,
          "eventHandlers": [
            {
              "global": false,
              "name": "@click",
              "actions": [
                {
                  "targetId": "dialog-0",
                  "name": "show",
                  "description": "Show dialog",
                  "argument": "undefined"
                },
                {
                  "targetId": "text-0",
                  "name": "setData",
                  "description": "Set text",
                  "argument": "'hallo'\n"
                }
              ]
            }
          ],
          "type": "ButtonView",
          "cid": "button-0"
        }
      ],
      "defaultValue": "={}",
      "type": "ContainerView",
      "eventHandlers": [],
      "cid": "container-0"
    },
    {
      "title": "",
      "content": {
        "tag": "p",
        "text": "",
        "type": "TextView",
        "eventHandlers": [
          {
            "global": false,
            "name": "@click",
            "actions": [
              {
                "targetId": "text-0",
                "name": "setData",
                "description": "Default action",
                "argument": "'hallo'\n"
              },
              {
                "targetId": "dialog-0",
                "name": "show",
                "description": "New action",
                "argument": "undefined"
              }
            ]
          }
        ],
        "cid": "text-0",
        "dataSource": "$parent"
      },
      "type": "DialogView",
      "eventHandlers": [],
      "cid": "dialog-0",
      "dataSource": "$parent"
    }
  ],
  "defaultValue": "={}",
  "type": "ContainerView",
  "eventHandlers": [],
  "cid": "container-1"
}
```

**Comment by renaudpawlak** [comment]
Note that I will also create an issue for your suggestion of having the ``cid`` appearing first in the serialized JSON.

**Comment by HFolkertH** [comment, marked as answer]
Hello Renaud,

Thanks very much for giving the answer!!!
I was already starting to doubt my sanity and IT skills
It works now

Regards
Folkert

--

**Reply by renaudpawlak** [reply]
You should never doubt ;)
Thank you for your help improving DLite!
